modules/Bio/EnsEMBL/Transcript.pm

```diff
@@ -2091,79 +2091,82 @@ sub get_all_translateable_Exons {
 
 sub translate {
   my ($self, $complete_codon) = @_;
+  my $codon_table_id;
 
   if ( !defined( $self->translation() ) ) { return undef }
 
-  my $mrna = $self->translateable_seq();
-
   # Alternative codon tables (such as the mitochondrial codon table)
   # can be specified for a sequence region via the seq_region_attrib
   # table.  A list of codon tables and their codes is at:
   # http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi
 
-  my $codon_table_id;
-  my ( $complete5, $complete3 );
   if ( defined( $self->slice() ) ) {
     my $attrib;
 
     ($attrib) = @{ $self->slice()->get_all_Attributes('codon_table') };
     if ( defined($attrib) ) {
       $codon_table_id = $attrib->value();
     }
+  }
 
-    ($attrib) = @{ $self->slice()->get_all_Attributes('complete5') };
-    if ( defined($attrib) ) {
-      $complete5 = $attrib->value();
-    }
+  $codon_table_id ||= 1;    # default vertebrate codon table
+  my $codon_table =
+    Bio::Tools::CodonTable->new( -id => $codon_table_id );
 
-    ($attrib) = @{ $self->slice()->get_all_Attributes('complete3') };
-    if ( defined($attrib) ) {
-      $complete3 = $attrib->value();
+  my $mrna = $self->translateable_seq();
+
+  my $delta = CORE::length($mrna) % 3;
+  if ( $delta > 0 ) {
+    # If we have a partial codon of we need to decide if we
+    # trim it or not to fix some bad behaviour in older bioperl
+    # versions
+    if ( $complete_codon ) {
+      # If we want to do the bad behavior of bioperl 1.6.1 and older
+      # where we guess the last codon if incomplete, pad an N
+      # to the mrna sequence
+      $mrna .= 'N' x (3 - $delta);
+    } else {
+      # Otherwise trim those last bp's off so the behavior is
+      # consistent across bioperl versions
+      chop $mrna for 1 .. $delta;
     }
   }
-  $codon_table_id ||= 1;    # default vertebrate codon table
+  
+  if ( CORE::length($mrna) < 1 ) { return undef }
```

Comment:
We need at least length 3 for the substr. While Perl will not complain, would this make sense?
If not, this should be "< 3" or even "< 6". Having start and stop codon be the same thing also does not make too much sense.

Reply:
What is returned on L2512
```
my $p_seq = $codon_table->translate($mrna, 0);
```
for an empty $mrna?
If it's an object, this is not equivalent to undef.

Reply:
Sorry. Upon thinking. We can have a partial codon that we can translate :)
And sometimes we don't have starts and stops. I assume the question if the model having just one AA derived from a partial codon is a valid one should be answered somewhere else. In the calling code, i.e. This bit is just for translating, imho. Up to you, though.

Reply:
OK, if this could make sense in some situations, leave it as-is.

Reply:
It makes no sense. But can work. Sorry.